src/AppInstallerCLICore/Workflows/ImportExportFlow.cpp

```diff
@@ -0,0 +1,204 @@
+        outputFileStream << packages;
+    }
+
+    void ReadImportFile(Execution::Context& context)
+    {
+        std::filesystem::path importFilePath{ context.Args.GetArg(Execution::Args::Type::ImportFile) };
+
+        // TODO: Handle errors
+        Json::Value jsonRoot;
+        std::ifstream{ importFilePath } >> jsonRoot;
+
+        auto packages = PackagesJson::ParseJson(jsonRoot);
+        context.Add<Execution::Data::PackageCollection>(packages);
+
+        if (packages.RequestsFromSources.empty())
+        {
+            AICLI_LOG(CLI, Warning, << "No packages to install");
+            AICLI_TERMINATE_CONTEXT(APPINSTALLER_CLI_ERROR_NO_APPLICATIONS_FOUND);
+        }
+    }
+
+    void SearchPackagesForImport(Execution::Context& context)
+    {
+        auto availableSources = Repository::GetSources();
+        std::vector<std::shared_ptr<IPackageVersion>> packagesToInstall;
+
+        // Aggregated source with all the required sources.
+        // We keep it as the source of the root context to keep all the
+        // source objects alive for install.
+        auto aggregatedSource = std::make_shared<Repository::CompositeSource>("*ImportSource");
+
+        // Look for the packages needed from each source independently.
+        // If a package is available from multiple sources, this ensures we will get it from the right one.
+        for (auto& requiredSource : context.Get<Execution::Data::PackageCollection>().RequestsFromSources)
+        {
+            if (!requiredSource.Details.Name.empty())
+            {
+                // For packages that come from a specific source, find the matching available source.
+                // Match required sources with the available sources by their arguments, as they may have been added with a different name.
+                // TODO: Can we do this with identifiers without opening the source?
+                std::optional<SourceDetails> matchingSource = {};
+                for (auto& availableSource : availableSources)
+                {
+                    if (availableSource.Arg == requiredSource.Details.Arg)
+                    {
+                        matchingSource = availableSource;
+                        break;
+                    }
+                }
+
+                if (matchingSource.has_value())
+                {
+                    requiredSource.Details.Name = matchingSource.value().Name;
+                }
+                else
+                {
+                    // TODO: Add option for ignoring/installing missing sources?
+                    AICLI_LOG(CLI, Error, << "Missing required source: " << requiredSource.Details.Name);
+                    AICLI_TERMINATE_CONTEXT(APPINSTALLER_CLI_ERROR_SOURCE_NAME_DOES_NOT_EXIST);
+                }
+            }
+
+            context << OpenNamedSource(requiredSource.Details.Name);
+            auto source = context.Get<Execution::Data::Source>();
+            aggregatedSource->AddAvailableSource(source);
```

Comment:
This pattern does not force the source to actually be used for a given package, which is what we want.  The goal should be that if a package should come from a specific source, then that source should be used exclusively.  Otherwise, the standard unnamed source (all combined) should be used.

Reply:
> This pattern does not force the source to actually be used for a given package

What do you mean?

Reply:
You are finding source 1, then adding it to the aggregate, then searching.  All of these packages will come from the correct source.

Then you find source 2, add it to the aggregate, and search.  These packages could be coming from source 1 or 2.

You should simply open the source and use it for the packages in question.  No need to use the aggregate.

Reply:
When I do the search, I'm not doing it in `context` which has the aggregate, but in `searchContext` which only has the individual source, so it should use only that one source unless I'm missing something.

I'm using the aggregate because some step of the installation (can't remember which atm) references the package's source so we need to keep it alive somehow (as the package has only a weak reference). Maybe keeping it as the root context's Source was not the best idea, more so since the composite source wasn't intended to be public. I can change it to keep around the sources as a list in the context.

But this all depends on me wanting to do all the searches before installing. If we change it to searching right before each install as you say, the aggregate can go away.